libp2p-xtra/src/listener.rs

```diff
@@ -0,0 +1,120 @@
+/// Should be used in conjunction with supervisor for continuous and resilient listening.
+pub struct Actor {
+    tasks: Tasks,
+    endpoint: Address<Endpoint>,
+    listen_address: Multiaddr,
+    /// Contains the reason we are stopping.
+    stop_reason: Option<Error>,
+}
+
+impl Actor {
+    pub fn new(endpoint: Address<Endpoint>, listen_address: Multiaddr) -> Self {
+        Self {
+            tasks: Tasks::default(),
+            endpoint,
+            listen_address,
+            stop_reason: None,
+        }
+    }
+
+    async fn check_connection_active_in_endpoint(&self) -> Result<()> {
+        let ConnectionStats {
+            listen_addresses, ..
+        } = self
+            .endpoint
+            .send(GetConnectionStats)
+            .await
+            .map_err(|_| Error::NoEndpoint)?;
+
+        listen_addresses
+            .contains(&self.listen_address)
+            .then(|| ())
+            .ok_or_else(|| anyhow!("Listener address not active in Endpoint"))
+    }
+}
+
+#[async_trait]
+impl xtra::Actor for Actor {
+    type Stop = Error;
+
+    async fn started(&mut self, ctx: &mut xtra::Context<Self>) {
+        let this = ctx.address().expect("we are alive");
+
+        let endpoint = self.endpoint.clone();
+        let listen_address = self.listen_address.clone();
+
+        {
+            let this = this.clone();
+            self.tasks.add_fallible(
+                endpoint
+                    .send(ListenOn(listen_address))
+                    .map_err(|e| Error::Failed { source: anyhow!(e) }),
+                move |e| async move {
+                    tracing::error!("Cannot listen on the endpoint");
+                    let _: Result<(), Disconnected> = this.send(e).await;
+                },
+            );
+        };
+
+        // Only start checking the connection after it had enough time to be established
+        tokio::time::sleep(CHECK_INTERVAL).await;
+        self.tasks
+            .add(this.send_interval(CHECK_INTERVAL, || CheckConnection));
+    }
+
+    async fn stopped(self) -> Self::Stop {
+        self.stop_reason.unwrap_or(Error::Unspecified)
+    }
+}
+
+#[xtra_productivity]
+impl Actor {
+    async fn handle(&mut self, msg: Error, ctx: &mut xtra::Context<Self>) {
+        self.stop_reason = Some(msg);
+        ctx.stop();
+    }
+
+    async fn handle(&mut self, _msg: CheckConnection, ctx: &mut xtra::Context<Self>) {
+        if self.check_connection_active_in_endpoint().await.is_err() {
+            self.stop_reason = Some(Error::ConnectionDropped);
+            ctx.stop();
+        }
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Listener failed")]
+    Failed { source: anyhow::Error },
+    #[error("Endpoint actor is disconnected")]
+    NoEndpoint,
+    #[error("Connection dropped from endpoint")]
+    ConnectionDropped,
+    #[error("Stop reason was not specified")]
+    Unspecified,
+}
+
+#[derive(Clone, Copy)]
+pub struct CheckConnection;
```

Comment:
🔧 I think these don't need to be `pub`.

Reply:
`Error` has to be public, for supervisor needs. The other one doesn't, fixed!